components/webext-storage/src/store.rs

```diff
@@ -110,18 +116,32 @@ impl Store {
     /// Returns the bytes in use for the specified items (which can be null,
     /// a string, or an array)
     pub fn get_bytes_in_use(&self, ext_id: &str, keys: JsonValue) -> Result<usize> {
-        api::get_bytes_in_use(&self.db, ext_id, keys)
+        let db = self.db.lock();
+        api::get_bytes_in_use(&db, ext_id, keys)
     }
 
     /// Returns a bridged sync engine for Desktop for this store.
-    pub fn bridged_engine(&self) -> sync::BridgedEngine<'_> {
+    pub fn bridged_engine(&self) -> sync::BridgedEngine {
         sync::BridgedEngine::new(&self.db)
     }
 
     /// Closes the store and its database connection. See the docs for
     /// `StorageDb::close` for more details on when this can fail.
-    pub fn close(self) -> result::Result<(), (Store, Error)> {
-        self.db.close().map_err(|(db, err)| (Store { db }, err))
+    pub fn close(self) -> Result<()> {
+        // Even though this consumes `self`, the fact we use an Arc<> means
+        // we can't guarantee we can actually consume the inner DB - so do
+        // the best we can.
+        let shared: SharedStorageDb = match Arc::try_unwrap(self.db) {
```

Comment:
thought: Could we unwrap this and panic if we didn't consume the inner DB? Cloning the `Arc` with `self.db.clone()` feels like a programmer bug that we'd want to catch early—I don't think an application could / should be expected to recover from `OtherConnectionReferencesExist`...

```suggestion
        let shared = Arc::try_unwrap(self.db).expect("Shouldn't have any other DB references");
```

Reply:
I was wondering how we would end up in this state too!

One possible (but unlikely?) case is that while we were closing the store, there was a sync operation still running that upgraded it's `Weak` reference into a strong one before the dropping :( I'm not sure however if that is prevented at a higher level in m-c though 🤔

Reply:
huh - that's a good point. I don't think anything will prevent this. I was thinking we could lock the mutex, but you can't into_inner() a locked mutex. hrmph

Reply:
FWIW this was the bases for my comment on the `Drop` implementation 😬  as long as the `Store` is dropped, eventually the `SharedStorageDb` will be dropped too! We can then close the underlying connection?

Reply:
but - to back up a bit, if `Arc::try_unwrap()` works, then we must have the only reference to the mutex, so it can't possibly be locked, right? eg, changing this to use `Mutex::get_mut()` also compiles

Reply:
Yup! If we can get the owned SharedStorageDb, we are guaranteed to have the only reference! I think it's the case where `try_unwrap` fails that's the problem :(

Reply:
But that doesn't change the underlying problem - a sync *might* be running as we hit this code, so it's that `Arc::try_unwrap()` which fails. We could take Lina's advice to panic in that situation, but I'm not sure how we can actually guarantee that never happens? ie, it's not what will guarantee an in-flight sync operation completes before we hit this code if we are syncing at shutdown?

Reply:
> ie, what will guarantee an in-flight sync operation completes before we hit this code if we are syncing at shutdown?

Unless there is synchronization at a higher level I don't think we can unless we add **additional synchronization** :(

We do know though, that unless the process is killed before the sync operation actually finishes, we'll **eventually** drop the inner `SharedStorageDb` because the `Store` is consumed (thus the only long-lasting strong reference). The problem becomes that the underlying SQL connection is not closed (i think?)

Reply:
> One possible (but unlikely?) case is that while we were closing the store, there was a sync operation still running that upgraded it's Weak reference into a strong one before the dropping :( I'm not sure however if that is prevented at a higher level in m-c though 🤔
>
> [...]
>
> what will guarantee an in-flight sync operation completes before we hit this code if we are syncing at shutdown?

Oooh, no, that's a great point! Desktop currently uses a single [task queue](https://searchfox.org/mozilla-central/rev/234ae43d3793289167926ef91bfb5145da91d4aa/xpcom/threads/TaskQueue.h#27-31) shared between the bridged engine and the store, which guarantees that all operations scheduled on it will execute serially, and which is why I thought we could panic here—but I don't think we can guarantee that for the other backends? 🤯

If you access the same `Store` from two different Swift dispatch queues (or a single concurrent dispatch queue), or two different Kotlin coroutine contexts, or even two different concurrently executing Rust threads...I think you're both totally right; it's very possible to end up with an in-flight sync operation that upgraded its `Weak` ref to a strong one running on one queue / thread, and a close operation running on a different queue, at the same time.

Reply:
I also think the [mutex that the bridge holds around the store](https://searchfox.org/mozilla-central/rev/234ae43d3793289167926ef91bfb5145da91d4aa/toolkit/components/extensions/storage/webext_storage_bridge/src/store.rs#45) helped here too. For something like places, I don't think the writer is ever explicitly closed 😬 (IIUC, Fenix treats the places api as a singleton and happily leaks it at shutdown)

Reply:
Yeah, definitely! I think you and @tarikeshaq are on the money—without some kind of additional synchronization to ensure that all `Store` calls execute serially, `Arc::try_unwrap` can fail.

But, it's okay for callers to ignore that, _as long as_ they're treating the `Store` as a singleton that's only closed at shutdown. We might leak the connection then, but the leak won't last for long, because the kernel will close all open file descriptors when the process goes away...whereas using `.expect()`/`.unwrap()` will cause a crash.

Does that sound right?

Reply:
That does sound right. The other part of the picture is avoiding the panic, which I think we can do by, as Tarik suggests, implementing `Drop` on `StorageDb` and explicitly doing the closing/forget on `writer`?

The main reason I think "this is fine" is because this problem already exists for all our other engines - including tabs, which is in desktop, and because desktop has `panic=abort`, this wouldn't go unnoticed - so I'm also interested to know when the connection close will panic in practice given I don't think we've ever seen evidence of it actually failing (which isn't to suggest we shouldn't pretend it can't though)

Reply:
> The main reason I think "this is fine" is because this problem already exists for all our other engines - including tabs, which is in desktop, and because desktop has `panic=abort`, this wouldn't go unnoticed...

For sure, and I'm wondering if Desktop avoids running into this today—and why we haven't seen evidence of it actually failing—because it's always using a single task queue. The store is making the assumption that it won't be used concurrently, and, because the task queue guarantees that there's only one in-flight call at a time, that assumption holds.